webrtc/README.md

```diff
@@ -0,0 +1,199 @@
+
+3. _A_ constructs _B_'s SDP offer locally based on _B_'s multiaddr and sets it
+   via
+   [`RTCPeerConnection.setRemoteDescription()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCPeerConnection/setRemoteDescription).
+
+4. _A_ establishes the connection to _B_. Yet to be determined how _B_ learns of
+   _A_ fingerprint. Potentially through A's STUN message USERNAME attribute. See
+   Open Questions below.
+
+5. _A_ initiates some authentication handshake _X_ to _B_ on a datachannel,
+   where _X_ allows _A_ and _B_ to authenticate each other's peer IDs. _X_ could
+   for example be Noise. See WebTransport specification as an example
+   https://github.com/libp2p/specs/pull/404. Still to be specified here for
+   WebRTC.
+
+6. On success of the authentication handshake _X_, the used datachannel is
+   closed and the plain WebRTC connection is used with its multiplexing
+   capabilities via datachannels.
+
+### Browser to Browser
+
+Scenario: Browser _A_ wants to connect to Browser node _B_ with the help of
+server node _R_.
+
+- Replace STUN with libp2p's identify and AutoNAT
+  - https://github.com/libp2p/specs/tree/master/identify
+  - https://github.com/libp2p/specs/blob/master/autonat/README.md
+- Replace TURN with libp2p's Circuit Relay v2
+  - https://github.com/libp2p/specs/blob/master/relay/circuit-v2.md
+
+## Open Questions
+
+_Where _Browser_ is replaced with the major desktop browsers of today (Chrome,
+Safari, Edge, Firefox)._
+
+- Should libp2p's WebRTC stack limit itself to using UDP only, or support WebRTC
+  on top of both UDP and TCP?
+
+### Direct connections
+
+#### Browser to public server node
+
+- Can _Browser_ generate a _valid_ SDP packet for the remote node based on the
+  remote's Multiaddr, where that Multiaddr contains the IP, UDP port and TLS
+  certificate fingerprint (e.g.
+  `/ip6/2001:db8::/udp/1234/webrtc/certhash/<hash>/p2p/<peer-id>`)? _Valid_ in
+  the sense that this generated SDP packet can then be used to establish a
+  WebRTC connection to the remote.
+
+- Do the major (Go / Rust / ...) WebRTC implementations allow us to accept a
+  WebRTC connection from a remote node without previously receiving an SDP
+  packet from such host?
+
+  How does the server learn the TLS certificate fingerprint of the browser? Is
+  embedding A's TLS certificate fingerprint in A's STUN message USERNAME
+  attribute a valid option?
+
+#### Browser to browser or non-public server node
+
+- Can _Browser_ know upfront its UDP port which it is listening for incoming
+  connections on? Does the browser reuse the UDP port across many WebRTC
+  connections? If that is the case one could connect to any public node, with
+  the remote telling the local node what port it is perceived on.
+
+- Can _Browser_ control the lifecycle of its local TLS certificate, i.e. can
+  _Browser_ use the same TLS certificate for multiple WebRTC connections?
+
+- Can two _Browsers_ exchange their SDP packets via a third server node using
```

Comment:
If the two browsers can securely exchange messages, they should ideally generate SDP with secure ICE credentials and exchange them followed by an exchange of trickled ICE candidates.

Reply:
> with secure ICE credentials

Today there is no way to exchange additional data, like ICE credentials, via the [DCUtR](https://github.com/libp2p/specs/blob/master/relay/DCUtR.md) protocol. That said, we could extend the protocol, in case there is no other way.

Reply:
We should consider not using DCUtR at all. DCUtR is built to exchange multiaddrs and to synchronize clocks to enable the hole punching. Here we need to exchange SDPs (which we might be able to reconstruct from multiaddrs, but we might as well skip that conversion step), and we don't need any clock synchronize clocks at all.

The alternative would be creating a new protocol that is run over relayed connections.

If I understand correctly, this protocol would need to support:
* exchange of the initial SDP
* exchange of trickle ICE candidates after that
* (not sure if there's a "that's all the ICE candidates I'll ever send" message)

Not sure about framing yet, my first impulse would be to define Protobuf messages to carry these, but there are probably other options.

Reply:
We could use the `onicegatheringstatechanged` callback to know when candidate gathering is completed. That being said, we could just wait for the gathering to complete and fetch the SDP from the `PeerConnection`. It should contain all the required candidates at this point.

Reply:
Documented `onicegatheringstatechanged` in  6c7f18e for now. Let's delay the discussion on candidate gathering, exchange and the usage of DCUtR to the second iteration of this specification. Resolving here.